tensorboard/webapp/metrics/views/main_view/card_grid_test.ts

```diff
@@ -0,0 +1,188 @@
+import {Component, Input} from '@angular/core';
+import {TestBed} from '@angular/core/testing';
+import {PluginType} from '../../data_source';
+import {By} from '@angular/platform-browser';
+import {NoopAnimationsModule} from '@angular/platform-browser/animations';
+import {CardLazyLoader, CardObserver} from '../card_renderer/card_lazy_loader';
+
+import {CardGridComponent} from './card_grid_component';
+import {CardGridContainer} from './card_grid_container';
+import {CardViewComponent} from '../card_renderer/card_view_component';
+import {CardViewContainer} from '../card_renderer/card_view_container';
+import {CardIdWithMetadata} from '../metrics_view_types';
+import {Store} from '@ngrx/store';
+import {MockStore, provideMockStore} from '@ngrx/store/testing';
+import {appStateFromMetricsState} from '../../testing';
+import {State} from '../../../app_state';
+import {selectors as settingsSelectors} from '../../../settings';
+import * as selectors from '../../../selectors';
+
+const scrollElementHeight = 100;
+
+// Utility funciton to allow use of await.
+function timeout(ms: number) {
+  return new Promise((resolve) => setTimeout(resolve, ms));
+}
+
+@Component({
+  selector: 'testing-component',
+  template: `
+    <div cdkScrollable>
+      <div class="placeholder">placeholder</div>
+      <metrics-card-grid
+        [cardIdsWithMetadata]="cardIdsWithMetadata"
+        [cardObserver]="cardObserver"
+        [groupName]="cardObserver"
+      ></metrics-card-grid>
+      <div class="placeholder">placeholder</div>
+    </div>
+  `,
+  styles: [
+    `
+      div {
+        position: fixed;
+        display: block;
+        height: ${scrollElementHeight}px;
+        overflow-y: scroll;
+      }
+      .placeholder {
+        position: relative;
+        height: 700px;
+        display: block;
+      }
+    `,
+  ],
+})
+class TestableComponent {
+  @Input() groupName: string = 'test group name';
+  @Input() cardIdsWithMetadata: CardIdWithMetadata[] = [
+    {
+      cardId: 'card1',
+      plugin: PluginType.SCALARS,
+      tag: 'tagA',
+      runId: null,
+    },
+    {
+      cardId: 'card2',
+      plugin: PluginType.SCALARS,
+      tag: 'tagA/Images',
+      runId: 'run1',
+      sample: 0,
+    },
+    {
+      cardId: 'card3',
+      plugin: PluginType.SCALARS,
+      tag: 'tagB/meow/cat',
+      runId: 'run1',
+      sample: 0,
+    },
+  ];
+  @Input() cardObserver: CardObserver = new CardObserver();
+}
+
+describe('card grid', () => {
+  let store: MockStore<State>;
+  beforeEach(async () => {
+    await TestBed.configureTestingModule({
+      imports: [NoopAnimationsModule, ScrollingModule],
+      declarations: [
+        CardGridComponent,
+        CardGridContainer,
+        TestableComponent,
+        CardViewComponent,
+        CardViewContainer,
+        CardLazyLoader,
+      ],
+      providers: [
+        provideMockStore({
+          initialState: appStateFromMetricsState(),
+        }),
```

Comment:
preferential thing: some test specify the mockStore but, in actually, you don't need this at all. Since we overrideSelector and make sure all states are mocked away (we don't want to deal with real reducer structure; i.e., reducer shape is an implementation detail that views should not be aware of and selectors should be the public API between reducer <--> view. In this world, view test knowing the shape is meh).

Reply:
I was thinking along those lines at one point. All selectors should be overridden so we don't actually need a state object. However, without this I get `NullInjectorError: No provider for Store!` is there a different Provider I can put here to resolve this issue? (feel free to message me offline to discuss.)

Reply:
You can provide a mock store but without the initialState.

Reply:
Ah I see what you mean. Works great. I had to override one more selector that I didn't even realize was being called but, all is working now. Thanks!